lib/submit-analytics.js

```diff
@@ -33,50 +104,184 @@ function sha256(data) {
     return hash.digest('hex');
 }
 
-if (isAnalyticsDisabled()) {
-    module.exports = function(){};
-} else {
+/**
+ * Send collected analytics data to Realm's servers over HTTPS
+ * @param  {[webhookProdDetails|webhookStageDetails|mixpanelDetails]} requestFactory Submission URL generator
+ * @param  {Object} payload analytics info
+ */
+async function dispatchAnalytics(requestFactory, payload) {
+    const https = require('https');
+
+    return new Promise((resolve, reject) => {
+        const webhookRequest = new requestFactory().buildRequest(payload);
+        var req = https.get(webhookRequest, function (result) {
+            result.on('data', (data) => {
+                // we don't log the data that is returned -- we're just interested in the header
+            });
+        });
+
+        req.on('response', response => {
+            response.on('end', () => {
+                resolve(
+                    {
+                        statusCode:  response.statusCode,
+                        statusMessage: response.statusMessage,
+                        errorMessage: '',
+                    });
+                return;
+            });
+
+            response.on('error', error => {
+                reject(
+                    {
+                        statusCode:  response.statusCode,
+                        statusMessage: response.statusMessage,
+                        errorMessage: error,
+                    });
+                return;
+            });
+        });
+    });
+}
+
+/**
+ * Collect analytics data from the runtime system
+ * @param  {Object} context contents of the app's package.json file
+ * @param  {string} eventName Custom tag that signifies what the app was doing when the analytics were submitted
+ * @returns {Object} Analytics payload
+ */
+function fetchPlatformData(context, eventName) {
     const os = require('os');
-    const request = require('request');
     const { machineId } = require("node-machine-id");
+    const environment = utils.getContext();
+
+    return new Promise((resolve, reject) => {
+        machineId()
+        .catch((e) => {
+            reject('Analytics submission failed');
```

Comment:
It's bad practice to reject a promise with a string for the same reason that one shouldn't throw a string. Although this is technically possible in JS, it won't have have a stack present, making debugging an error much more complicated.
```suggestion
            const err = new Error('Analytics submission failed');
            reject(err);
```

Reply:
If you moved this `.catch` after your call to `.then` any errors thrown from computing the payload would also be caught by this block and return a meaningful error.

Reply:
Actually, this error is mostly for debugging purposes.  In production, it should fail silently.  Would it be more appropriate to resolve() the promise in the .catch()?